### PR TITLE
Fix position misalignment when past_key_values is externally built

### PR DESCRIPTION
**Pull request by Piggy343288**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47597)

## Problem

`generate()` accepts a user-built `past_key_values`, but `_prefill()` silently mishandles it when the `attention_mask` length differs from the `input_ids` length — exactly the case when prefill and decoding are done separately:

```python
# Step 1: prefill
out = model(**inputs)           # input_ids=[1, 1800], mask=[1, 1800]

# Step 2: compress / rebuild cache
new_cache = DynamicCache()
for li in range(n_layers):
    new_cache.update(k_compressed, v_compressed, li)

# Step 3: generate — position tracking is broken
model.generate(
    input_ids=last_token,       # [1, 1]
    attention_mask=old_mask,    # [1, 1800]
    past_key_values=new_cache,
)
```

This produces **wrong output with no exception or crash** — the kind of bug that passes CI and makes it into production.

On Qwen3-VL-8B (MMBench-Video, 1747 QA pairs), the same INT8 quantization applied post-fill scores 0.196 judge accuracy vs 0.303 baseline (−10.7 pp), while the identical quantization applied inside the model's forward pass via hooks scores 0.295 (−0.8 pp). The 12× penalty gap comes entirely from the `generate()` path, not from quantization error.

## Root cause

`_prefill()` at line 3919:

```python
if attention_mask is not None and input_ids.shape[1] == attention_mask.shape[1]:
    next_sequence_length = input_ids.shape[1] - past_length
```

When `past_key_values` was built externally and `attention_mask` covers the full prefill history while `input_ids` has only the continuation token, this condition is `False` → `next_sequence_length = None`. `prepare_inputs_for_generation` then cannot correctly compute `cache_position` and `position_ids`, breaking the position chain from the first decode step.

## Changes

**1. Core fix — `_prefill()` (`generation/utils.py`)**

Add an `elif` branch: when `past_length > 0`, `attention_mask` covers cached history, but `input_ids` has only new tokens, compute `next_sequence_length` from the new token count.

**2. API — `DynamicCache.from_kv_tensors()` (`cache_utils.py`)**

New classmethod providing a canonical entry point for external cache construction. Backward-compatible, no overhead.

**3. Warning — `_prefill()`**

Log warning_once when `past_key_values` has cached tokens but `attention_mask` is missing entirely.

## Who can review?

@Cyrilvallez @ArthurZucker

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
Please see the comments/doc on cache continuation:
```
# When restarting from previous cache, the `input_ids` are either the FULL sequence, including previous inputs,
# or only the new tokens but in this case the attention_mask still contains the FULL sequence (because otherwise we may
# lose some early padding tokens information). So slice inputs according to that if needed
# When restarting from `inputs_embeds`, it's always the FULL sequence, and we always need to slice
```
